docs/new_flux_ideas_and_guide.md

```diff
@@ -0,0 +1,985 @@
+| coyote_creek | CA | 51 | 2020-02-22T16:05 | 2020-02-22T17:00 |
+| santa_monica| CA | 48 | 2020-02-22T17:01 | 2020-02-22T18:00 |
+| santa_monica| CA | 48 | 2020-02-22T17:31 | 2020-02-22T18:00 |
+| santa_monica| CA | 48 | 2020-02-22T17:46 | 2020-02-22T18:00 |
+| puget_sound | WA | 40 | 2020-02-22T17:10 | 2020-02-22T18:00 |
+| puget_sound | WA | 40 | 2020-02-22T17:40 | 2020-02-22T18:00 |
+
+What this does is for each 1 hour period, it groups the rows by state and then selects the rows with the max value. 
+Because we included the `windowColumn` argument, that is included in the result with the ending timestamp of the 
+window that each record fell into.
+
+Since so many of the records have tied `max(bottom_degrees)` values, we see that many of the original 
+rows were selected in the result. We can also see the `window` that is inserted to specify 
+which block of time the selector was working across.
+
+The following functions can be used in the **Selector Predicate Block**: `first`, `last`, `min`, `max`, `top`, `bottom`
+
+#### Calling Aggregates without the Aggregate function
+The `aggregate` function provides a shorthand for computing multiple aggregates at a time. However, the 
+functions passed in can also be called on tables. For example:
+
+```
+// get the most recently written record
+@my_data.h2o_temperature |> last()
+
+// get the most recently written record for each location
+@my_data.h2o_temperature |> last(by: ["location"])
+
+// get the average for each hour for each location
+@my_data.h2o_temperature{time > -3h} |> mean(by: ["location"], window: 1h)
+``` 
+
+Selectors can only be called with the `select` function.
+
+#### Count, Distinct, and CountDistinct
+The `count` function is an aggregate while the `distinct` function is a selector. The `countDistinct` function is 
+a helper function that will combine operations for you.
+ 
+```
+@my_data.h2o_temperature{state, time > -3h}
+    |> select(fn: distinct(state), window: 1h, windowColumn: "time")
+```
+
+Would return the following result
+
+| state | time
+| --- | --- |
+| CA | 2020-02-22T16:00
+| CA | 2020-02-22T17:00
+| CA | 2020-02-22T18:00
+| WA | 2020-02-22T18:00
+
+We can see that CA reported in all three hours while both CA and WA reported in only the last hour. Also see that 
+distinct discards columns that aren't in the `by` argument or the `distinct` column itself. So we use the `time` 
+name for where we put the window markers for each distinct state. You could then count the number of states reporting 
+per hour:
+
+```
+@my_data.h2o_temperature{state, time > -3h}
+    |> select(fn: distinct(state), window: 1h)
+    |> timeShift(-1m)
+    |> count("state", window: 1h)
+```
+
+Would produce:
+
+| state | time
+| --- | --- |
+| 1 | 2020-02-22T16:00
+| 1 | 2020-02-22T17:00
+| 2 | 2020-02-22T18:00
+
+See that we had to use the `timeShift` function to move the timestamps of the windows back so that the resulting 
```

Comment:
Also, not clear why I’d need `timeShift`? Feels magical to have to know you need to move times...

Reply:
The use case where `timeShift` comes up is wanting to compare data from different times. For example say I wanted to plot the average temperature for each 1h for the entire day. I would have a plot with 24 points. Now say I wanted to put yesterdays data on the same plot?

How would you like to describe that query in Flux?


This is how I would expect it to be done:

```
@data.temperature{surface_degrees, location, time > -24h}
    |> mean(by:["location"], window: 1h)

@data.temperature{surface_degrees, location, time > -48h and time < -24h}
    |> mean(by:["location"], window: 1h)
    |> timeShift(1h)
```

Now my plot has both yesterday's and today's data overlaid. 

Open to other paradigms for approaching this use case.

Reply:
hmm, is that `timeShift` suppose to be `timeShift(24h)`... this is interesting, although I don't know that I'd want to use it for this use case... UI concerns are a UI responsibility in my view...  looking at the transformed data, I think you'd have to use a yield to use this programmatically, does that sound about right?

Reply:
@jsteenb2 edited the post, yeah it was supposed to be `timeShift(24h)`.

The use case is not unique to wanting to visualize data. For example I want to create an alert where today's average is significantly different than yesterday's average at the for the same  time of day.

```
today = @data.temperature{surface_degrees, location, time > -24h}
    |> mean(by:["location"], window: 1h)

yesterday = @data.temperature{surface_degrees, location, time > -48h and time < -24h}
    |> mean(by:["location"], window: 1h)
    |> timeShift(24h)

today.surface_degrees - yesterday.surface_degrees
   |> monitor.check(...)
```

Yeah I am ignoring how data is yielded in these examples as its not really related to the use case of `timeShift`

Reply:
right on, this makes a lot more sense. I'm still not a fan of the syntax of measurement minus a measurement. Feels awkward, but not a hill for me to die on, just a friendly concern 😀

Reply:
I agree. this doesn't make sense to me.